url: urlexclude/allow commands; auto-title ignores URLs in ANY other command

Exirel opened the pull request:
### Description

Close #1870

And because nothing is simple: I also added aliases for `urlpban` and `urlpunban` to ban/unban patterns instead of URL directly.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

cottongin commented:
I like this, my only suggestion is what do you think about renaming the commands to use a different word than "ban"?

At least to me using "ban" implies banning someone from using the plugin. I know it also includes "url" in the command name, so perhaps it is less confusing to others, but the plugin is also named `url.py`... (maybe a better suggestion would be to rename the plugin `title.py` or something)

Exirel commented:
If you have something better than `ban`, I'll take it. `urlexclude` maybe?

cottongin commented:
Yes sure, exclude makes more sense to me

dgw commented on sopel/modules/url.py:
```diff
@@ -153,6 +153,96 @@ def shutdown(bot):
+        # escape the URL to ensure a valid pattern
+        url = re.escape(url)
+
+    patterns = bot.settings.url.exclude or []
```
No need to fall back on an empty list; `ListAttribute` already presents as empty list if not set.

```suggestion
    patterns = bot.settings.url.exclude
```

dgw commented on sopel/modules/url.py:
```diff
@@ -153,6 +153,96 @@ def shutdown(bot):
+
+    # update settings
+    patterns.append(url)
+    bot.settings.url.exclude = patterns
```
Not necessary if `patterns` is a simple assignment to `bot.config.url.exclude`; it'll be the same object reference.

```suggestion
```

dgw commented on sopel/modules/url.py:
```diff
@@ -153,6 +153,96 @@ def shutdown(bot):
+        # escape the URL to ensure a valid pattern
+        url = re.escape(url)
+
+    patterns = bot.settings.url.exclude or []
```
See comments/suggestions in `url_ban()` above.

```suggestion
    patterns = bot.settings.url.exclude
```

dgw commented on sopel/modules/url.py:
```diff
@@ -153,6 +153,96 @@ def shutdown(bot):
+
+    # update settings
+    patterns.remove(url)
+    bot.settings.url.exclude = patterns
```
See `url_ban()` comments.

```suggestion
```

dgw reviewed (changes requested):
"urlpunban" lol 😹

I couldn't take the time just now to think about the `# re-compile` steps, but surely there's a way to manage `bot.memory['url_exclude']` without running a whole list comprehension every time?

Exirel commented:
> I couldn't take the time just now to think about the `# re-compile` steps, but surely there's a way to manage `bot.memory['url_exclude']` without running a whole list comprehension every time?

I think there is, but I couldn't bother to do something complicated without the proper unit-test to back me up. I think the tradeoff is good for now. I know I tend to be a bit lazy with tests when it comes to plugin, because I never know if I really want to invest in them right now, or wait for when they are all external plugins. After all, I did a 100% test coverage on sopel-help, and I'm doing that for remind too...

dgw reviewed (dismissed):
Squash at will, sir. 🖖

Exirel commented:
OK, I squashed and rebased after some further fix and changes, making it worthy of a second full review, I think.

I've settled for `urlexclude` and `urlallow` in hope of making @cottongin happy. We discussed this with @dgw but neither of us had an opinion about it. We decided to keep the `url(pun)ban` because it's funny.

dgw reviewed (approved):
Retitled to reflect the extra fix Exi sneakily added, so I remember to mention it in the changelog.
